### DISCUSSION TITLE
[RFC] Improving auth solutions in Expo 🤔

### DISCUSSION THREAD

**Discussion by EvanBacon** [expo/expo#7789] [Proposals and Ideas]
Currently `expo-app-auth` has a bunch of different problems, Google auth is confusing due to the two different packages, and we often recommend people don't use `expo-auth-session` for production apps. 

How can we fix this and try to prevent it from happening again in the future?

Related: 
- https://github.com/expo/expo/issues/6370 
- [AuthSession on Expo web](https://expo.canny.io/feature-requests/p/authsession-on-expo-web)

# Issues

- None of the auth modules support web. 
- Because everything except AuthSession is native, bug fixes take a while to propagate. 
- It's unclear how you should build a provider specific abstraction on top of Expo. 
- Test coverage is really poor.
- You sometimes need to be an expert in how OAuth works if you want to get these working with broken auth providers like Reddit or Github.

# Goals

- Frequent updates and faster bug fixes by using more Expo primitives like Crypto, Random, SecureStore, WebBrowser, Linking.
- Universal support for authentication.
- Better managed and bare workflow support.
- Improve testing for auth.
- Reduce app size by consolidating auth solutions as much as possible.
- Easier use of auth.expo.io with any auth provider.

# Background 

Auth in Expo is split across a few modules, I'll explain what they do but in this RFC I mostly want to focus on `expo-auth-session` and `expo-app-auth`. 

### expo-auth-session

- **workflow**: Mostly managed.
- **provider**: Any OAuth provider. 
- **platforms**: iOS, Android, partial web

A fully JS module that uses a proxy service (auth.expo.io) to make setup easier in managed projects. 

#### Cons

- Does not have any protection against [Cross-Site Request Forgery](https://tools.ietf.org/html/rfc6749#section-10.12).
- No support for PKCE.
- No support for auto discovery or OpenID Connect.
- Requires a deep understanding of auth to craft the authentication URI.
- Awkward to swap out for a better solution in the future.
- Message on iOS says "open in expo" and there is an interstitial page.

### expo-app-auth

- **workflow**: any
- **provider**: any
- **platforms**: iOS, Android

Native binding to AppAuth-iOS and AppAuth-Android for performing OAuth with any provider. It protects against [Cross-Site Request Forgery](https://tools.ietf.org/html/rfc6749#section-10.12), supports PKCE, and OpenID Connect.

#### Cons

- iOS and Android native packages work differently
- Native modules are not very flexible, making it difficult to support non spec-compliant auth providers like Github or Reddit.
- Duplicate native code: All of the features are already implemented elsewhere in Expo packages.
- The auth method combines two steps together: auth and code exchange. This is default on iOS, and emulated on Android. When combined, the end-user has far less control over the login process, the errors are also somewhat cryptic because you're not completely sure which step failed.
- No support for dynamic client registration.
- Doesn't work the same as other Expo modules, it opts to throwing errors when events like "dismiss" or "cancel" take place.
- Linking API is awkward and doesn't account for switching between workflows.
- Doesn't support Expo proxy for quick bootstrapping.

### expo-google-app-auth

- **workflow**: Same as `expo-app-auth`
- **provider**: Google
- **platforms**: Same as `expo-app-auth`

Extends `expo-app-auth` providing a simple interface for Google web auth.

#### Cons

- Because it's part of expo/expo, it's harder to update often. 
- It requires native code via `expo-app-auth`.

### expo-google-sign-in

- **workflow**: bare-only
- **provider**: Google
- **platforms**: iOS, Android

Native binding to `GoogleSignIn` that provides extra functionality on Android. 

#### Cons

- The iOS functionality is akin to a more limited version of `expo-google-app-auth` as it only works in the bare-workflow.
- Includes an extra native package and native binding.

### expo-facebook

- **workflow**: any
- **provider**: Facebook
- **platforms**: iOS, Android

Native binding to Facebook SDK with ability to deep link into native FB app for authentication.

#### Cons

- Since Facebook's auth API is pretty limited to configure we require the native package to get optimal performance.
- Native method attempts to link into the Facebook app for auth, this isn't part of the OAuth spec and therefore harder to replicate. Even if it could be replicated it's probably not work it for the overhead.

# Proposal

Initially I thought rewriting `expo-app-auth` (#7450) would be the best solution but I think we could do better. Instead I think we should add full support for OAuth and OpenID Connect to `expo-auth-session`, migrate `expo-google-app-auth` to using `expo-auth-session`, and then deprecate `expo-app-auth`. In time I think we could also replace the iOS section of `expo-google-sign-in` with `expo-auth-session` which would add managed workflow support via the Expo proxy.

We would build PKCE support using `expo-crypto` and `expo-random`, perform authentication requests via `expo-web-browser`, and construct workflow appropriate URIs with expo Linking.
- We could extend `expo-auth-session` this API using packages like `expo-secure-store`, and the (tbd) cookies library.
- Full JS means full mocking, test coverage could be better than ever before.
- Easier to build community driven abstractions on top of.
- Unified functionality across iOS, Android, and web. 
  - Because mobile-web requires all loading to happen before a user interaction occurs it would mean that you'd have to build part of your app differently to account for just mobile-web. If we unify the API then we can ensure auth works the same everywhere.

## Why Auth Session?

- Breaking changes are a big concern at Expo, since AuthSession doesn't implement much we'd be closer to starting fresh by implementing the changes there. 
- AuthSession is also already a fully JS package, changes could be pushed OTA whenever we want.
- You can use the Expo proxy with AuthSession already.
- It'd be much easier to migrate from using the proxy to using a proper redirect URI.
- We'd be free to do whatever we want with the API which is important for supporting web and other platforms in the future.

**Comment by tonycoco** [comment]
A few things I ran into:

1. What deterred me was the lack of clear examples or samples for any kind of usage. Need a kitchen sink app where a reference solution can be clearly defined. Mainly, how React Navigation can be used and how a simple login/register/forgot pass flow can be coded.
2. I had my own login flow in an existing web application and wanted to simply expose some API endpoints and get the two integrated. There weren't many options to see as to how to do this, I'm guessing, obvious starting point and use-case.
3. Expo also needs also support some of the big out-of-the-box Auth providers for people that either have those integrations already or are starting from zero. Also, abstraction on this layer would allow anyone to create their own auth "plugin" with some boilerplate code (thus, sort of fixing #2 in the process).

**Reply by EvanBacon** [reply]
> 1

These are great points! Integration guides with react-navigation is an interesting idea I hadn't thought of before, I'll talk with the team about this!

We've historically only supported the particularly difficult auth methods with their own modules. i.e Facebook, Google, and Apple. 

I've been working on a list of [different auth examples](https://github.com/expo/expo/blob/@evanbacon/auth-session/auth-request/docs/pages/versions/unversioned/sdk/auth-session.md#examples) (currently 12) using the new API. All will be tested in their entirety using our [kitchen sink auth example](apps/native-component-list/src/screens/AuthSessionScreen.tsx).

> 2

Could you elaborate on what part you needed clarification on. There are two main APIs involved with auth `fetch`, and `expo-web-browser`.

> 3

I think maybe it'll make sense to create community abstractions in the future which have stronger types, more visual setup guides, secure rehydration, and requests for the corresponding API.

A big pain-point in setting up auth is the URI redirects, to make bare-workflow setup easier I've created `npx uri-scheme` which makes configuration really simple. I should document it more in the package tho.

**Comment by MarcusSorealheis** [comment]
I’m just getting started with Expo. I think that an Okta integration and an Authy integration would serve the community well and help adoption and contribution to accelerate.

**Comment by desnor** [comment]
> * You can use the Expo proxy with AuthSession already.

Would we still have the option of using the default redirect pattern that we currently have in `expo-app-auth` where we don't have to go through the `auth.expo.io` proxy as part of the oauth flow?

**Reply by EvanBacon** [reply]
Yes, but now you can toggle the proxy on/off with a single boolean - effectively making development much faster. Also you'll still be able to test services like Google and Facebook from the client without needing to add a custom redirect!

**Comment by EvanBacon** [comment]
Excited to announce the first release of this new auth proposal https://twitter.com/Baconbrix/status/1256985914749759488

**Comment by martin-bucinskas** [comment]
@EvanBacon you mentioned that `expo-app-auth` supports PKCE, is this enabled by default?
I cannot find any reference to documentation on this. It would be great if you could point me in the right direction.

I feel like I'm missing a trick here. We have a managed Expo app and we're trying to authenticate the user against an OIDC identity provider (in our case it's Cognito).

It seems to two options we have are either AuthSession or AppAuth.
It seems that AuthSession will support a PKCE flow, but doesn't have a mechanism for refreshing the token. 

AppAuth, on the other hand, does have a mechanism for refreshing the token but doesn't seem to support a PKCE flow.

I don't want to increase the token timeout because that's a security risk. I also don't want to use an implicit flow because that's also a security risk.

There is an [open PR](https://github.com/expo/expo/pull/8051) against AuthSession to add refresh token support, but the fact that the library has been out for this long without it makes me wonder if I'm missing something.
